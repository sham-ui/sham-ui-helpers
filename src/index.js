import { this$ } from 'sham-ui-macro/ref.macro';

/**
 * Add this$.setField helper to component
 * @param {Function} options Component options
 */
export function SetField( options ) {
    const state = options();
    this$.setField = field => value => state[ field ] = value;
}

/**
 * Add this$.toggleField helper to component
 * @param {Function} options Component options
 */
export function ToggleField( options ) {
    const state = options();
    this$.toggleField = field => () => state[ field ] = !state[ field ];
}
