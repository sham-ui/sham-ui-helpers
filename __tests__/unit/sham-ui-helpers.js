import { onclick, oninput } from 'sham-ui-directives';
import renderer, { compileAsSFC } from 'sham-ui-test-helpers';
import { ToggleField, SetField } from '../../src/index';

it( 'toggleField', () => {
    const meta = renderer(
        compileAsSFC( {
            ToggleField,
            Toggler: compileAsSFC`
                <template>
                    <button :onclick={{onClick}}>Toggle!</button>
                </template>
                
                <script>
                    function Toggler( options ) {
                        options( {
                            [ $.onClick ]() {}
                        } );
                    }
                
                    export default Component( Template, Toggler );
                </script>
            `
        } )`
            <template>
                Toggled: {{toggled}}
                <Toggler
                    onClick={{this$.toggleField( $.toggled )}}
                />
            </template>
            
            <script>
                function Dummy( options ) {
                    options( {
                        [ $.toggled ]: false
                    } );
                }

                export default Component( ToggleField, Template, Dummy );
            </script>
        `,
        {},
        {
            directives: {
                onclick
            }
        }
    );

    meta.ctx.container.querySelector( 'button' ).click();
    expect( meta.toJSON() ).toMatchSnapshot();
    expect( meta.component.options.toggled ).toBe( true );

    meta.ctx.container.querySelector( 'button' ).click();
    expect( meta.toJSON() ).toMatchSnapshot();
    expect( meta.component.options.toggled ).toBe( false );
} );

it( 'setField', () => {
    const meta = renderer(
        compileAsSFC( {
            SetField,
            Input: compileAsSFC`
                <template>
                    <input :oninput={{this$.handleInput}} />
                </template>
                
                <script>
                    function Input( options ) {
                        const onChanged = $();
                
                        const state = options( {
                            [ onChanged ]() {}
                        } );
                
                        this$.handleInput = e => state[ onChanged ]( e.target.value );
                    }
                
                    export default Component( Template, Input );
                </script>
            `
        } )`
            <template>
                Value: {{value}}
                <Input
                    onChanged={{this$.setField( $.value )}}
                />
            </template>
            
            <script>
                function Dummy( options ) {
                    options( {
                        [ $.value ]: ''
                    } );
                }

                export default Component( SetField, Template, Dummy );
            </script>
        `,
        {},
        {
            directives: {
                oninput
            }
        }
    );

    function fill( value ) {
        const element = meta.ctx.container.querySelector( 'input' );
        element.value = value;
        element.dispatchEvent( new Event( 'input' ) );
    }

    fill( 'Hi!' );
    expect( meta.toJSON() ).toMatchSnapshot();
    expect( meta.component.options.value ).toBe( 'Hi!' );

    fill( 'Hello!' );
    expect( meta.toJSON() ).toMatchSnapshot();
    expect( meta.component.options.value ).toBe( 'Hello!' );
} );
