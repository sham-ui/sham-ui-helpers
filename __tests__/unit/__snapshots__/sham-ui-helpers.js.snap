// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setField 1`] = `
Object {
  "Options": Object {
    "value": "Hi!",
  },
  "html": "
 Value: Hi!
 <input><!--0-->
",
}
`;

exports[`setField 2`] = `
Object {
  "Options": Object {
    "value": "Hello!",
  },
  "html": "
 Value: Hello!
 <input><!--0-->
",
}
`;

exports[`toggleField 1`] = `
Object {
  "Options": Object {
    "toggled": true,
  },
  "html": " Toggled: true<button>Toggle!</button><!--0-->",
}
`;

exports[`toggleField 2`] = `
Object {
  "Options": Object {
    "toggled": false,
  },
  "html": " Toggled: false<button>Toggle!</button><!--0-->",
}
`;
